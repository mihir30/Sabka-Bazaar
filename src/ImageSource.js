import { Image } from "./components/atoms";
const ImageSource = ({ alt, source, className }) => {
  return (
    <Image source={require(`${source}`)} alt={alt} className={className} />
  );
};

export default ImageSource;
