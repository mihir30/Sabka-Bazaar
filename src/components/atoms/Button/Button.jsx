import React from "react";
import "./Button.scss";

export default function Button({
  children,
  buttonType = "",
  method, dataTestId, 
  ...props
}) {
  return (
    <button className={"btn-" + buttonType} onClick={method} {...props} data-testid={dataTestId}>
      {children}
    </button>
  );
}
