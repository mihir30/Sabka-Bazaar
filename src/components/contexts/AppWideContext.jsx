import React, { createContext, useReducer } from "react";
import { reducer } from "../reducers/ReducerContext";

const initialState = {
  count: 0,
  products: {},
};

export const AppWideContext = createContext(initialState);

const AppWideContextProvider = ({ children }) => {
  const [cartItems, dispatch] = useReducer(reducer, initialState);

  return (
    <AppWideContext.Provider value={{ cartItems, dispatch }}>
      {children}
    </AppWideContext.Provider>
  );
};

export default AppWideContextProvider;
