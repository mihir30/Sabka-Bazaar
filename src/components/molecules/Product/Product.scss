@mixin align-center() {
    text-align: center;
    vertical-align: middle;
}

@mixin border-style() {
    border: 2px solid black;
}

.products {
    @include border-style();
    margin: 110px 10px 10px;
}

.btn-BuyNow {
    display: none;
}

.btn-buynow-tab-phone {
    display: block;
    padding: 5px 5px;
    background: #cc0044;
    color: white;
    border-radius: 2px;
    font-weight: 700;
    margin-top: 30px;
    font-size: 15px;
    height: 35px;
    width: 210px;
    border-color: #cc0044;
}

.price {
    display: none;
}

.products .product-heading {
    padding: 10px 0 0 20px;
}

.products-heading {
    padding-left: 3%;
}

.product-description {
    overflow-y: scroll;
    padding: 5px;
    width: 100%;
    height: 80px;
    border-radius: 5px;
    text-align: justify;
    background: #e7e5e5de;
    padding: 2px;
    font-weight: 550;
}

.product-description::-webkit-scrollbar {
    display: none;
}

.item {

    text-align: start;
    text-decoration: none;
    font-size: 20px;
    height: 75px;
}

.product {
    padding: 2px;
    height: 100%;
    display: inline-block;
    border-radius: 5px;
    box-shadow: 5px 0px 10px 0px #9e9e9e;
    border-right: 1px solid rgba(0, 0, 0, 0.5);
    border-bottom: 1px dashed rgba(0, 0, 0, 0.5);
    margin: 0px 15px 25px 15px;
}

.img-desc-btn {
    display: flex;
}

.product-description {
    display: block;
}

.product-images {
    width: 125px;
    height: 150px;
}

.text-box {
    margin-left: 5px;
}

@media (min-width: 768px) {
    .btn-buynow-tab-phone {
        margin-top: 10px;
        margin-left: -100px;
        width: 295px;
        text-align: center;
    }

    .product-images {
        width: 100px;
        height: 100px;
    }

    .product-description {
        height: 100px;
        margin: 0px -3px;
    }

    .btn-buynow-tab-phone {
        margin-top: 10px;
    }

    .product-images {
        height: 110px;
    }
}

@media (min-width: 1024px) {
    .img-desc-btn {
        display: block;
    }

    .product-images {
        width: 250px;
        height: 250px;
    }

    .price {
        display: block;
        margin-top: 10px;
        margin-right: 60px;
        font-size: 18px;
        font-weight: 500;
    }

    .btn-buynow-tab-phone {
        display: none;
    }

    .btn-BuyNow {
        display: block;
        padding: 10px 20px;
        background: #cc0044;
        color: white;
        border-radius: 2px;
        font-weight: 700;
        font-size: 15px;
        border-color: #cc0044;
    }

    .price-and-buy-now-btn {
        display: flex;
        justify-content: center;
        margin-top: 15px;
    }
}