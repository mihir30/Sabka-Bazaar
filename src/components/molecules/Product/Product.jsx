import React, { useContext } from "react";
import "./Product.scss";
import { AppWideContext } from "../../contexts/AppWideContext";
import { Button } from "../../atoms";
import ImageSource from "../../../ImageSource";
import { toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
toast.configure();
const Product = ({ name, imageUrl, price, stock, text, id }) => {
  const {
    dispatch,
    cartItems: { products },
  } = useContext(AppWideContext);

  function addItemToCart() {
    if (!products[id]) {
      toast.success(`${name} has been added to your cart`, {
        theme: "light",
        position: toast.POSITION.BOTTOM_LEFT,
        autoClose: 2500,
      });
      dispatch({
        type: "ADD_ITEM",
        product: {
          [id]: {
            id,
            imageUrl,
            name,
            price,
            stock,
            quantity: 1,
          },
        },
      });
    } else {
      toast.success(`${name} has been added to your cart again`, {
        theme: "light",
        position: toast.POSITION.BOTTOM_LEFT,
        autoClose: 2500,
      });
      dispatch({
        type: "EDIT_ITEM",
        id: id,
      });
    }
  }

  return (
    <div className="product" id={`product${id}`}>
      <h2 className="item">{name}</h2>
      <div className="img-desc-btn">
        <div className="image-box">
          <div className={`images${id}`}>
            <ImageSource
              source={imageUrl}
              alt={`Image ${id}`}
              className="product-images"
              id={`thumbnail${id}`}
            />
          </div>
        </div>
        <div className="text-box" data-testid="product-card-component">
          <div className="product-description">{text}</div>
          <div className="price-and-buy-now-btn">
            <h3 className="price" id={`productPrice${id}`}>
              MRP Rs.{price}
            </h3>

            <Button
              buttonType="BuyNow"
              children="Buy Now"
              method={() => addItemToCart()} data-testid="buy-now-btn"
            ></Button>
          </div>
          <div>
            <Button
              buttonType="buynow-tab-phone"
              children={`Buy Now @ MRP Rs.${price}`}
              method={() => addItemToCart()} data-testid="buy-now-btn"
            ></Button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Product;