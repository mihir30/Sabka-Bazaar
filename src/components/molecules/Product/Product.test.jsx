import { render } from "@testing-library/react";
import React from "react";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import ImageSource from "../../../ImageSource";
import { AppWideContext } from "../../contexts/AppWideContext";
import Button from '../../atoms/Button/Button';
import Product from "./Product";

describe("Product Component", () => {
  const addItemToCart = jest.fn();
  jest.doMock("../../../ImageSource", () => {
    return {
      __esModule: true,
      // the "default export"
      default: () => <ImageSource source="./static/images/products/fruit-n-veg/kiwi-green.jpg" alt={"Image 5b6c6a7f01a7c38429530883"} className={"product-images"} />
    }
  });

  jest.doMock("../../atoms/Button/Button", () => {
    return {
      __esModule: true,
      // the "default export"
      default: () => {
        <Button
          buttonType="BuyNow"
          children="Buy Now"
          method={() => addItemToCart()} data-testid="buy-now-btn"
        ></Button>
      }
    }
  });

  jest.doMock("../../atoms/Button/Button", () => {
    return {
      __esModule: true,
      // the "default export"
      default: () => {
        <Button
          buttonType="buynow-tab-phone"
          children={`Buy Now @ MRP Rs.87`}
          method={() => addItemToCart()} data-testid="buy-now-btn"
        ></Button>
      }
    }
  });


  it("should render the comp", () => {
    render(<AppWideContext.Provider value={{
      dispatchEvent,
      cartItems: {
        products: {
          id: "5b6c6a7f01a7c38429530883", imageUrl: "./static/images/products/fruit-n-veg/kiwi-green.jpg",
          name: "Fresho Kiwi - Green, 3 pcs", price: 87, stock: 50,
          quantity: 1,
        }
      }
    }}><BrowserRouter><Product id={"5b6c6a7f01a7c38429530883"} imageUrl={"./static/images/products/fruit-n-veg/kiwi-green.jpg"}
      name={"Fresho Kiwi - Green, 3 pcs"} price={87} stock={50} text={"Some dummy text"} /></BrowserRouter></AppWideContext.Provider>)
  });

  it("should give the right product component snapshot", () => {
    const tree = renderer.create(<AppWideContext.Provider value={{
      dispatchEvent,
      cartItems: {
        products: {
          id: "5b6c6a7f01a7c38429530883", imageUrl: "./static/images/products/fruit-n-veg/kiwi-green.jpg",
          name: "Fresho Kiwi - Green, 3 pcs", price: 87, stock: 50,
          quantity: 1,
        }
      }
    }}><BrowserRouter><Product id={"5b6c6a7f01a7c38429530883"} imageUrl={"./static/images/products/fruit-n-veg/kiwi-green.jpg"}
      name={"Fresho Kiwi - Green, 3 pcs"} price={87} stock={50} text={"Some dummy text"} /></BrowserRouter></AppWideContext.Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  })


});