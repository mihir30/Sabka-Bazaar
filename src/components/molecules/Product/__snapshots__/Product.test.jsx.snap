// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Component should give the right product component snapshot 1`] = `
<div
  className="product"
  id="product5b6c6a7f01a7c38429530883"
>
  <h2
    className="item"
  >
    Fresho Kiwi - Green, 3 pcs
  </h2>
  <div
    className="img-desc-btn"
  >
    <div
      className="image-box"
    >
      <div
        className="images5b6c6a7f01a7c38429530883"
      >
        <img
          alt="Image 5b6c6a7f01a7c38429530883"
          className="product-images"
          data-testid="image"
          src="kiwi-green.jpg"
        />
      </div>
    </div>
    <div
      className="text-box"
      data-testid="product-card-component"
    >
      <div
        className="product-description"
      >
        Some dummy text
      </div>
      <div
        className="price-and-buy-now-btn"
      >
        <h3
          className="price"
          id="productPrice5b6c6a7f01a7c38429530883"
        >
          MRP Rs.
          87
        </h3>
        <button
          className="btn-BuyNow"
          onClick={[Function]}
        >
          Buy Now
        </button>
      </div>
      <div>
        <button
          className="btn-buynow-tab-phone"
          onClick={[Function]}
        >
          Buy Now @ MRP Rs.87
        </button>
      </div>
    </div>
  </div>
</div>
`;
